### DISCUSSION TITLE
Consider removing Dynaconf

### DISCUSSION THREAD

**Discussion by noklam** [kedro-org/kedro#3877] [Idea]
![image](https://github.com/kedro-org/kedro/assets/18221871/eb375aec-a739-46b4-b558-af6b6de8f2df)

## Extend Discussion
> Settings management (default value, template, validation, parsing)
Doesn't seem to be particular useful as overlap with configloader already. Validation these days are not really useful, there are probably better choices. The only thing that I recalled is custom config loader validation but 0.19 there is only `OmegaConfigLoader` and we expect most project stick with it. If the validation is useful, maybe we should consider validating datasets etc?

> Fully support of environment to override existing config
? `KEDRO_LOGGING_CONFIG`?

> Optional layer environment
- Kedro's ConfigLoader (base, local, prod etc)

> Secret Vault & Credential Management
Seems like the most interesting part for Kedro

> CLI
Unknown, the dynaconf init creates a set of files that are very different from Kedro. `dynaconf write py -v key=value` is a way to update settings.py dynamically.

Dynaconf comes with many dependencies and currently we are only using it for schema validation.

Alternative:
Actually leverage features of dyanconf like dotenv native support, credentials management etc

More:
### Settings https://www.dynaconf.com/settings_files/
> .py - Not Recommended but supported for backwards compatibility.

**Comment by noklam** [comment]
Some historical context:

[Refactor the way Kedro loads settings.py](https://github.com/McK-Private/private-kedro/pull/975) (private link)

Refactor the way Kedro loads settings and make it similar to Django (?)

Here's a list of links for inspiration
1. https://docs.djangoproject.com/en/dev/topics/settings/#using-settings-without-setting-django-settings-module
3. https://docs.djangoproject.com/en/dev/topics/settings/#using-settings-in-python-code
5. https://docs.djangoproject.com/en/dev/topics/settings/#default-settings
7. https://github.com/django/django/blob/master/django/conf/global_settings.py
9. [https://github.com/django/django/blob/master/django/conf/_init_.py#L96](https://github.com/django/django/blob/master/django/conf/__init__.py#L96)
11. [https://github.com/django/django/blob/master/django/conf/_init_.py#L131](https://github.com/django/django/blob/master/django/conf/__init__.py#L131)

Double-check that this works with multiprocessing on Windows, i.e. spawn multiprocessing

Example resulting API
```
from kedro.framework.project import settings, configure_project
configure_project(package_name) to be called in all places immediately after _add_src_to_path
```

- All plugins, Kedro code and user code can access settings by settings.CONF_ROOT, etc.
- No need to pass package_name to KedroSession and KedroContext
- Make the change backwards compatible and issue a deprecation warning for `package_name`, saying that it is no longer needed and will be removed from 0.18

**Reply by astrojuanlu** [reply]
I'll be very honest, I was never a fan of `configure_project` and its aunt `bootstrap_project`. With all our talk about not using `global` variables [in our training materials](https://github.com/kedro-org/kedro-academy/tree/main/iswe4dx), I've always found it weird that we seem to get a pass:

https://github.com/kedro-org/kedro/blob/a53500afe5cd9739362b066e24771cd26b0c63bb/kedro/framework/project/__init__.py#L293-L294

**Reply by astrojuanlu** [reply]
In any case, the critical thing here is probably `settings.<VAR>` access, I guess?

**Comment by astrojuanlu** [comment]
If we ever wanted native support for Hashicorp (or OpenBao https://github.com/openbao/openbao) we could do that ourselves, possibly through a plugin:

https://github.com/dynaconf/dynaconf/blob/0177b68a14036f32adc06f3e3596d31f945c31aa/dynaconf/loaders/vault_loader.py#L14-L15

**Comment by noklam** [comment]
![image](https://github.com/kedro-org/kedro/assets/18221871/d6098e3e-b1d6-4d77-b469-9b091abeee3e)
Found this today
